### ISSUE TITLE
SchemaDumper unable to generate schema for schema.rb file

### ISSUE THREAD

**Issue by DebbieGillespie** [rsim/oracle-enhanced#802]
When we run `./bin/rails db:migrate`, the tables are created in our Oracle DB, but the schema.rb file contains the following errors:

```
# Could not dump table "ar_internal_metadata" because of following NoMethodError
#   undefined method `[]' for nil:NilClass

# Could not dump table "tags" because of following NoMethodError
#   undefined method `[]' for nil:NilClass
```

Our migration file is as follows:

```ruby
class CreateTags < ActiveRecord::Migration[5.0]
  def change
    create_table :tags, id: false do |t|
      t.string :name
      t.integer :aplus_id
      t.timestamps
    end

    add_index :tags, :aplus_id, unique: true
  end
end
```

We know that the nil is coming from the `@types` reference on line 141, essentially the following block of code in schema_dumper.rb:
```ruby
            column_specs = columns.map do |column|
              raise StandardError, "Unknown type '#{column.sql_type}' for column '#{column.name}'" if @types[column.type].nil?
              next if column.name == pk
              @connection.column_spec(column, @types)
            end.compact
```

The @types was removed from active record in https://github.com/rails/rails/commit/9e4ed2f996d8dd7eaac0399f4efb5ce58aa593f8. It looks like they moved to using a valid_type? method in https://github.com/rails/rails/commit/c2e203173221991e104a9eb872025788b9fc4763.

**Comment by davinlagerroos** [comment]
As a workaround, we replaced `@types` on [L151](https://github.com/rsim/oracle-enhanced/blob/5c1f2079aadd467e58c603878cb0d571b60fd386/lib/active_record/connection_adapters/oracle_enhanced/schema_dumper.rb#L151) and [L153](https://github.com/rsim/oracle-enhanced/blob/5c1f2079aadd467e58c603878cb0d571b60fd386/lib/active_record/connection_adapters/oracle_enhanced/schema_dumper.rb#L153) with the old implementation, `@connection.native_database_types`. That gave us to a new error:

```
# Could not dump table "ar_internal_metadata" because of following ActiveRecord::ConnectionAdapters::OracleEnhancedConnectionException
#   "DESC " failed; does it exist?

# Could not dump table "tags" because of following ActiveRecord::ConnectionAdapters::OracleEnhancedConnectionException
#   "DESC " failed; does it exist?
```

**Comment by davinlagerroos** [comment]
Looks the description error is due to `@table_name` being nil in the columns generated on [shema_dumper#L118](https://github.com/rsim/oracle-enhanced/blob/5c1f2079aadd467e58c603878cb0d571b60fd386/lib/active_record/connection_adapters/oracle_enhanced/schema_dumper.rb#L118). 

This due to the call to `super` in `OracleEnhancedColumn#initialize`. As of [0aa83f3b8](https://github.com/rails/rails/commit/0aa83f3b88c71df4f29b8c52d6297e2e5886e258) `ActiveRecord::ConnectionAdapters::Column#initialize` sets `@table_name` to nil. ... and they set it later on with instance_variable_set. As of [6b6a01c](https://github.com/rails/rails/commit/6b6a01c52bacb2df2914029309144eacff25a3de) `table_name` is an argument to `ActiveRecord::ConnectionAdapters::Column#initialize` but that is only on master, not the beta release of ActiveRecord 5. 

Seems like the best answer would be to pass `table_name` in the call to `super`, but that is not an option until master gets released. Another option would be to move the setting of `@table_name` to after the `super` call. 

The latter allows me to generate a schema.db file. However, now all the columns have `{default => table name}` i.e.:

```
ActiveRecord::Schema.define(version: 20160406164156) do

  create_table "ar_internal_metadata", id: false, force: :cascade do |t|
    t.string    "key",                  default: -> { "ar_internal_metadata" }
    t.string    "value",                default: -> { "ar_internal_metadata" }
    t.timestamp "created_at", limit: 6, default: -> { "ar_internal_metadata" }, null: false
    t.timestamp "updated_at", limit: 6, default: -> { "ar_internal_metadata" }, null: false
  end

  create_table "tags", id: false, force: :cascade do |t|
    t.string    "name",                                default: -> { "tags" }
    t.integer   "aplus_id",             precision: 38, default: -> { "tags" }
    t.timestamp "created_at", limit: 6,                default: -> { "tags" }, null: false
    t.timestamp "updated_at", limit: 6,                default: -> { "tags" }, null: false
    t.integer   "parent_id",            precision: 38, default: -> { "tags" }
  end

  add_index "tags", ["aplus_id"], name: "index_tags_on_aplus_id", unique: true

end
```

**Comment by davinlagerroos** [comment]
...Or maybe the defaults was a fluke. I tried migrating and dumping again and the defaults are no longer showing up in my schema.rb file.

**Comment by yahonda** [comment]
Closed as #810 merged to rails5 branch.
